File: cpg-language-llvm/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/cpp/CompilationDB.java
/*
 * Copyright (c) 2021, Fraunhofer AISEC. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 *                    $$$$$$\  $$$$$$$\   $$$$$$\
 *                   $$  __$$\ $$  __$$\ $$  __$$\
 *                   $$ /  \__|$$ |  $$ |$$ /  \__|
 *                   $$ |      $$$$$$$  |$$ |$$$$\
 *                   $$ |      $$  ____/ $$ |\_$$ |
 *                   $$ |  $$\ $$ |      $$ |  $$ |
 *                   \$$$$$   |$$ |      \$$$$$   |
 *                    \______/ \__|       \______/
 *
 */
package de.fraunhofer.aisec.cpg.frontends.cpp;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.LinkedList;
import java.util.List;
import java.util.stream.Collectors;

public class CompilationDB {

  public static List<String> getIncludeFIlesInTheFolders(List<String> includeFilesDirectories) {
    List<String> includeFiles = new LinkedList<>();
    includeFilesDirectories.forEach(
        item -> {
          try {
            List<String> filesInFolder =
                Files.walk(Paths.get(item))
                    .filter(Files::isRegularFile)
                    .map(Path::toString)
                    .collect(Collectors.toList());
            includeFiles.addAll(filesInFolder);
          } catch (IOException e) {
            e.printStackTrace();
          }
        });
    return includeFiles;
  }

  public static List<String> getIncludeDirectories(String stringVal) {
    if (stringVal == null || stringVal.length() == 0) {
      return null;
    }
    // get all the -I flag files
    List<String> words = List.of(stringVal.split(" "));
    List<String> includeFilesDirectories = new LinkedList<>();
    for (String word : words) {
      if (word.startsWith("-I")) {
        includeFilesDirectories.add(word.substring(2)); // adds the directory excluding the -I field
      }
    }
    return includeFilesDirectories;
  }

  public static List<String> getIncludeDirectories(List<String> commandVals) {
    //     ['clang', 'main.c', '-o', 'main.c.o'],
    // The I vals come after -I
    List<String> includeFilesDirectories = new LinkedList<>();
    for (int i = 0; i < commandVals.size(); i++) {
      if (commandVals.get(i) != null && commandVals.get(i).startsWith("-I")) {
        if (i + 1 != commandVals.size()) {
          includeFilesDirectories.add(commandVals.get(i + 1));
        }
      }
    }
    return includeFilesDirectories;
  }
}

File: cpg-core/src/main/java/de/fraunhofer/aisec/cpg/frontends/cpp/CompilationDB.kt
/*
 * Copyright (c) 2021, Fraunhofer AISEC. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 *                    $$$$$$\  $$$$$$$\   $$$$$$\
 *                   $$  __$$\ $$  __$$\ $$  __$$\
 *                   $$ /  \__|$$ |  $$ |$$ /  \__|
 *                   $$ |      $$$$$$$  |$$ |$$$$\
 *                   $$ |      $$  ____/ $$ |\_$$ |
 *                   $$ |  $$\ $$ |      $$ |  $$ |
 *                   \$$$$$   |$$ |      \$$$$$   |
 *                    \______/ \__|       \______/
 *
 */
package de.fraunhofer.aisec.cpg.frontends.cpp

import org.slf4j.LoggerFactory
import java.util.*

/** This class is used to extract the details inside a compilation database. */
object CompilationDB {
    private val log = LoggerFactory.getLogger(CompilationDB::class.java)

    /**
     * Gets the include directories the from the string value provided. Example for a compile
     * commdand is : "/usr/local/bin/g++-7 -I/Users/me/prj/Calendar/calendars -g -std=c++11 -o
     * calendar_run.dir/main.cpp.o -c /Users/me/prj/Calendar/main.cpp"
     *
     * This method returns the include-paths in the above the command.
     */
    fun parseIncludeDirectories(stringVal: String): List<String>? {
        if (stringVal == null || stringVal.isEmpty()) {
            return null
        }
        // get all the -I flag files
        val words = java.util.List.of(*stringVal.split(" ").toTypedArray())
        val includeFilesDirectories: MutableList<String> = LinkedList()
        for (word in words) {
            if (word.startsWith("-I")) {
                includeFilesDirectories.add(
                    word.substring(2)
                ) // adds the directory excluding the -I field
            }
        }
        return includeFilesDirectories
    }

    /**
     * Gets the include directories the from the array value provided. Example for a compile
     * commdand is : ['clang', 'main.c', '-o', 'main.c.o'] This method returns the include-paths in
     * the above command.
     */
    fun parseIncludeDirectories(commandVals: List<String>): List<String>? {
        //     ['clang', 'main.c', '-o', 'main.c.o'],
        // The I vals come after -I
        if (commandVals == null || commandVals.isEmpty()) {
            return null
        }
        val includeFilesDirectories: MutableList<String> = LinkedList()
        for (i in commandVals.indices) {
            if (commandVals[i] != null && commandVals[i]!!.startsWith("-I")) {
                if (i + 1 != commandVals.size) {
                    includeFilesDirectories.add(commandVals[i + 1])
                }
            }
        }
        return includeFilesDirectories
    }
}
